Collect class property types as "used symbols"

martinssipenko opened the pull request:
Added collection of class property types and symbols of use statements.

Ocramius commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorFunctionalTest.php:
```diff
@@ -128,6 +129,26 @@ public function testWontCollectAnyUsageTypes(): void
+        $this->traverseStringAST('<?php class Foo { public My\ReturnType $foo; }');
+
+        self::assertSameCollectedSymbols(
+            ['My\ReturnType'],
```
`My\PropertyType` perhaps?

Ocramius commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorFunctionalTest.php:
```diff
@@ -128,6 +129,26 @@ public function testWontCollectAnyUsageTypes(): void
+
+    public function testWillCollectUseTypes(): void
+    {
+        $this->traverseStringAST('<?php use My\ReturnType;');
```
IMO this is not to be implemented.

The problem is that `use` statements can be partial symbols.

For example, in this case, `Doctrine\ORM\Mapping` is not a defined symbol:

```php
use Doctrine\ORM\Mapping as ORM;

// ...

/** @ORM\Entity() */
class MyEntity {
    // ...
```

I use this very often also with functions:

```php
use Psl\Type;

// ...

Type\object(MyClass:class)
    ->assert($value);
```

martinssipenko replied:
Fair enough, I'll remove the "use" part from this PR, but could you perhaps give me some hints on what would be the best way to implement symbols used in annotations?

Ocramius replied:
> but could you perhaps give me some hints on what would be the best way to implement symbols used in annotations?

Perhaps we can do that for PHP 8.1 attributes? Annotations are not that problematic anymore, and people will slowly move away from them...

martinssipenko replied:
I think they are still used quite a lot to tell static analyzers what types to infer, for example:
```php
/**
 * @var array<\Ramsey\Uuid>
 */
private array $uuids;
```

In such a case if ramsey/uuid is not installed the checker should fail, no?

Ocramius replied:
Ah, that stuff: there are tools for that in https://github.com/phpDocumentor/ReflectionDocBlock (which allows you to avoid re-inventing the wheel).

martinssipenko replied:
Ok, I'll see what I can do about it in a separate PR. Thanks!

Ocramius reviewed (approved):
LGTM, thanks @martinssipenko!

Ocramius commented:
Note: I have to mark this as `BC Break`, because upgrading to a version with this new collector will lead to failures downstream (behavioral change).

I'll re-target this for `4.0.x`, but don't have a branch yet, so merge needs some holding off :)

martinssipenko commented:
Maybe we can find a way to release these sorts of changes without introducing a new major each time behavior changes? The approach used by PHPStan with bleeding-edge seemed to work quite well. It would allow end-users to opt-in into all the latest checks that could be rolled out as enabled by default in the next major.

Ocramius commented:
> without introducing a new major each time behavior changes? 

Meh, it's better to just declare a new major release each time, and have downstream upgrade to it when they want/can.

Upgrades to new majors are supposed to be exactly for that that.

PHPStan had to invent its own mechanism because it did not have a sensible versioning mechanism for a long while.

Ocramius commented:
@maglnet can we perhaps release `4.0.0` next, with this specific BC break in it? :thinking:

maglnet commented:
@Ocramius for me a `4.0.0` would be fine.

Ocramius commented:
Alright, I'll rename the last `3.x` branch to `4.0.0`, and delete the next minor milestone. Only bugfixes allowed on `3.x` after that move :)

Ocramius reviewed (changes requested):
@martinssipenko I've rebased your patch, and mutation tests now fail due to new un-tested mutations.

Can you please check the diffs around the files you changed? It indicates that either:

 * tests have to be added
 * code has to be simplified

martinssipenko commented:
@Ocramius sure thing.

Ocramius commented on src/ComposerRequireChecker/NodeVisitor/UsedSymbolCollector.php:
```diff
@@ -52,6 +52,7 @@ public function enterNode(Node $node)
         $this->recordFunctionReturnTypeUsage($node);
         $this->recordConstantFetchUsage($node);
         $this->recordTraitUsage($node);
+        $this->recordProperyTypeUsage($node);
```
This is the problematic code path:

```diff
5) /home/runner/work/ComposerRequireChecker/ComposerRequireChecker/src/ComposerRequireChecker/NodeVisitor/UsedSymbolCollector.php:55    [M] MethodCallRemoval

--- Original
+++ New
@@ @@
         $this->recordFunctionReturnTypeUsage($node);
         $this->recordConstantFetchUsage($node);
         $this->recordTraitUsage($node);
-        $this->recordProperyTypeUsage($node);
+        
         return parent::enterNode($node);
     }
     private function recordExtendsUsage(Node $node) : void
```

This path seems to not be tested at all, with current unit tests (integration tests have `@coversNothing`, hence a unit test is needed)

martinssipenko replied:
Out of interest, why do integration tests have `@coversNothing`?

Ocramius replied:
Integration tests "leak" coverage, as in they provide partial coverage for a very wide surface area, but are supposed to be treated as "smoke test" for the whole thing, rather than in-depth testing.

In mutation tests, they lead to extremely slow mutation test runs (slow test applied to a lot of generated mutants => slowness multiplied) and therefore it is quite important to exclude them, favouring instead smaller and more efficient unit tests.

martinssipenko commented:
I've added a test for the change that I introduced. If I run infection locally it still reports mutations around my change to be escaping, however, if I mutate my code manually and run tests then tests fail.

Ocramius commented:
Let's please avoid merging upstream (like in bc34a96 ) - rebases instead :D

martinssipenko commented:
I somehow have managed to totally fuck up the history of this PR. @Ocramius Should I just create a new one?

localheinz commented on src/ComposerRequireChecker/NodeVisitor/UsedSymbolCollector.php:
```diff
@@ -52,6 +52,7 @@ public function enterNode(Node $node)
         $this->recordFunctionReturnTypeUsage($node);
         $this->recordConstantFetchUsage($node);
         $this->recordTraitUsage($node);
+        $this->recordProperyTypeUsage($node);
```
```suggestion
        $this->recordPropertyTypeUsage($node);
```

localheinz commented on src/ComposerRequireChecker/NodeVisitor/UsedSymbolCollector.php:
```diff
@@ -196,9 +197,22 @@ private function recordTraitUsage(Node $node): void
         }
     }
 
+    private function recordProperyTypeUsage(Node $node): void
```
```suggestion
    private function recordPropertyTypeUsage(Node $node): void
```

localheinz commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorTest.php:
```diff
@@ -289,4 +291,37 @@ public function testBeforeTraverseResetsRecordedSymbols(): void
         $this->assertCount(0, $this->visitor->getCollectedSymbols());
     }
+
+    public function testProperyType(): void
```
```suggestion
    public function testPropertyType(): void
```

Ocramius commented:
Haha, happens, especially when learning to rebase :D 

It's fine, I'll see if I can extract only the relevant commits, and force-push :)

localheinz commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorTest.php:
```diff
@@ -289,4 +291,37 @@ public function testBeforeTraverseResetsRecordedSymbols(): void
+        $this->assertContains('Bar', $symbols);
+    }
+
+    public function testIgnoresNonNameProprtyType(): void
```
```suggestion
    public function testIgnoresNonNamePropertyType(): void
```

localheinz commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorTest.php:
```diff
@@ -289,4 +291,37 @@ public function testBeforeTraverseResetsRecordedSymbols(): void
+        $this->assertCount(0, $symbols);
+    }
+
+    public function testIgnoresUnhadledNodeTypes(): void
```
```suggestion
    public function testIgnoresUnhandledNodeTypes(): void
```

martinssipenko committed:
Collect property type usage

martinssipenko committed:
Fix typo in method name

martinssipenko commented:
@Ocramius history fixed 😓

Ocramius commented:
Ah, alright - was trying it out on my end, but you were faster :muscle:

localheinz commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorTest.php:
```diff
@@ -289,4 +291,37 @@ public function testBeforeTraverseResetsRecordedSymbols(): void
+
+        $this->visitor->enterNode($node);
+
+        $symbols = $this->visitor->getCollectedSymbols();
+        $this->assertCount(1, $symbols);
+        $this->assertContains('Bar', $symbols);
```
```suggestion
        $symbols = $this->visitor->getCollectedSymbols();
                
        $this->assertCount(1, $symbols);
        $this->assertContains('Bar', $symbols);
```

Ocramius replied:
Is this only about spacing? If so, then let's leave it as is, and propose new CS rules in a separate patch.

localheinz commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorTest.php:
```diff
@@ -289,4 +291,37 @@ public function testBeforeTraverseResetsRecordedSymbols(): void
+
+        $this->visitor->enterNode($node);
+
+        $symbols = $this->visitor->getCollectedSymbols();
+        $this->assertCount(0, $symbols);
```
```suggestion
        $symbols = $this->visitor->getCollectedSymbols();

        $this->assertCount(0, $symbols);
```

localheinz commented on test/ComposerRequireCheckerTest/NodeVisitor/UsedSymbolCollectorTest.php:
```diff
@@ -289,4 +291,37 @@ public function testBeforeTraverseResetsRecordedSymbols(): void
+
+        $this->visitor->enterNode($node);
+
+        $symbols = $this->visitor->getCollectedSymbols();
+        $this->assertCount(0, $symbols);
```
```suggestion
        $symbols = $this->visitor->getCollectedSymbols();

        $this->assertCount(0, $symbols);
```

Ocramius commented:
Thanks @martinssipenko! I'll roll a new release once current `4.0.x` patches are merged

martinssipenko commented:
Thanks!
